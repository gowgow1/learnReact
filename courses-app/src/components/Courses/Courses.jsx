import { useState } from 'react';

import CourseCard from './components/CourseCard';
import SearchBar from './components/SearchBar';
import Button from '../common/Button';

import './index.css';

const Courses = ({ togglePage, initialCourseList, initialAuthors }) => {
	const courseList = initialCourseList.map((course) => {
		const { id, title, description, creationDate, duration, authors } = course;
		const autorNames = authors?.map(
			(id) => initialAuthors?.find((item) => item?.id === id)?.name
		);

		return (
			<CourseCard
				key={id}
				title={title}
				description={description}
				author={autorNames}
				duration={duration}
				created={creationDate}
			/>
		);
	});

	const [courses, setCoursesList] = useState(courseList);
	const [InputValue, setInputValue] = useState('');

	const searchAbility = ({ target: { value } }) => {
		setInputValue(value);

		if (InputValue.length <= 1) {
			setCoursesList(courseList);
		}
	};

	const searchElements = () => {
		setCoursesList(
			courseList.filter(
				({ props: { title }, key }) =>
					title.toLowerCase().includes(InputValue.toLowerCase()) ||
					(key + '').includes(InputValue)
			)
		);
	};

	return (
		<div className='Courses-wrap'>
			<div className='settings'>
				<SearchBar
					searchAbility={searchAbility}
					searchElements={searchElements}
				/>
				<Button text='Add new course' onClick={togglePage} />
			</div>
			<div className='cards-wrap'>{courses}</div>
		</div>
	);
};

export default Courses;
