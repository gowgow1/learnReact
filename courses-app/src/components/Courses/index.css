.Courses-wrap {
	padding: 15px 20px;
	border: 2px solid lightskyblue;
}
.settings {
	display: flex;
	row-gap: 5px;
	flex-wrap: wrap;
	justify-content: space-between;
	column-gap: 20px;
	margin-bottom: 15px;
}
.settings .button,.input{
	padding: 5px 10px;
	border-radius: 0;
}