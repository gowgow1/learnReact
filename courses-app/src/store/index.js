import { createStore, combineReducers } from 'redux';
import courses from './courses/reducer';
import user from './user/reducer';

const store = createStore(combineReducers({ courses, user }));

export default store;
