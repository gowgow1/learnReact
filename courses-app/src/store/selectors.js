export const getUser = ({ user }) => user;
export const getCourses = ({ courses }) => courses;
export const getAuthors = ({ authors }) => authors;
