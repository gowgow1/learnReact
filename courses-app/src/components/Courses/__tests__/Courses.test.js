import { Provider } from 'react-redux';
import * as reactRedux from 'react-redux';
import { BrowserRouter, useHistory } from 'react-router-dom';
import { render, fireEvent, screen } from '@testing-library/react';

import Courses from '../Courses';
import * as selectors from '../../../store/selectors';
import { mockedAuthorsList, mockedCoursesList } from '../../../constants';

jest.mock('react-router-dom', () => ({
	...jest.requireActual('react-router-dom'),
	useHistory: jest.fn(),
}));

jest.mock('../../../store/selectors', () => ({
	getCourses: jest.fn(),
	getAuthors: jest.fn(),
	getUser: jest.fn(),
}));

jest.spyOn(reactRedux, 'useDispatch');

const mockedState = {
	user: {
		isAuth: true,
		name: 'Test Name',
		token: 'ddd32432-34324sfs-few2113',
		role: 'admin',
		email: 'test@test.ru',
	},
	courses: mockedCoursesList,
	authors: mockedAuthorsList,
};

const mockedStore = {
	getState: () => mockedState,
	subscribe: jest.fn(),
	dispatch: jest.fn(),
};

beforeEach(() => {
	selectors.getAuthors.mockReturnValue(mockedState.authors);
	selectors.getCourses.mockReturnValue(mockedState.courses);
	selectors.getUser.mockReturnValue(mockedState.user);
});

afterEach(() => {
	selectors.getAuthors.mockClear();
	selectors.getCourses.mockClear();
	selectors.getUser.mockClear();
});

describe('Courses', () => {
	it('Courses rendered', () => {
		const { container } = render(
			<Provider store={mockedStore}>
				<BrowserRouter>
					<Courses />
				</BrowserRouter>
			</Provider>
		);
		expect(container).toMatchSnapshot();
	});

	it('displays all courses from courses array', () => {
		const { getAllByTestId } = render(
			<Provider store={mockedStore}>
				<BrowserRouter>
					<Courses />
				</BrowserRouter>
			</Provider>
		);
		expect(getAllByTestId('card')).toHaveLength(mockedCoursesList.length);
	});

	//I dont know but mutcher 'toBeEmptyDOMElement' doesn't work from this library
	// it('should display Empty container if courses array length is 0', () => {
	// 	selectors.getCourses.mockReturnValue([]);
	// 	const { getByTestId } = render(
	// 		<Provider store={mockedStore}>
	// 			<BrowserRouter>
	// 				<Courses />
	// 			</BrowserRouter>
	// 		</Provider>
	// 	);
	// 	expect(getByTestId('cards-wrapper')).toBeEmptyDOMElement();
	// });

	it('should be redirect to courseAdd after click on button', () => {
		const historyMock = jest.fn();
		useHistory.mockReturnValue({
			location: {
				pathname: 'http://localhost:3000/courses/add',
			},
			push: historyMock,
		});
		render(
			<Provider store={mockedStore}>
				<BrowserRouter>
					<Courses />
				</BrowserRouter>
			</Provider>
		);
		const add = screen.getByTestId('add');
		fireEvent.click(add);
		expect(historyMock).toHaveBeenCalledWith('/courses/add');
	});
});
