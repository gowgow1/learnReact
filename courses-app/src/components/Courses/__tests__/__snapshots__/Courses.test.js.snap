// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Courses Courses rendered 1`] = `
<div>
  <div
    class="Courses-wrap"
  >
    <div
      class="settings"
    >
      <div
        class="search-bar"
      >
        <div
          class="input-wrap"
        >
          <label
            for="search"
          />
          <input
            class="input"
            id="search"
            placeholder="Enter course name..."
            type="text"
            value=""
          />
        </div>
        <button
          class="button"
        >
          Search
        </button>
      </div>
      <button
        class="button"
        data-testid="add"
      >
        Add new course
      </button>
    </div>
    <div
      class="cards-wrap"
      data-testid="cards-wrapper"
    >
      <div
        class="card"
        data-testid="card"
      >
        <div
          class="card-content"
        >
          <h2
            class="title"
          >
            JavaScript
          </h2>
          <div
            class="description"
          >
            Lorem Ipsum is simply dummy text of the printing and
            typesetting industry. Lorem Ipsum
            has been the industry's standard dummy text ever since the
            1500s, when an unknown
            printer took a galley of type and scrambled it to make a type
            specimen book. It has survived
            not only five centuries, but also the leap into electronictypesetting, remaining essentially u
            nchanged.
          </div>
        </div>
        <div
          class="card-info"
        >
          <div
            class="author"
          >
            <strong>
              Authors: 
            </strong>
            Vasiliy Dobkin, Nicolas Kim
          </div>
          <div
            class="duration"
          >
            Duration: 
            02:40
          </div>
          <div
            class="created"
          >
            Created: 
            8/3/2021
          </div>
          <div
            class="btn-wrap"
          >
            <button
              class="button"
            >
              Show course
            </button>
            <button
              class="button"
            >
              E
            </button>
            <button
              class="button"
            >
              D
            </button>
          </div>
        </div>
      </div>
      <div
        class="card"
        data-testid="card"
      >
        <div
          class="card-content"
        >
          <h2
            class="title"
          >
            Angular
          </h2>
          <div
            class="description"
          >
            Lorem Ipsum is simply dummy text of the printing and
            typesetting industry. Lorem Ipsum
            has been the industry's standard dummy text ever since the
            1500s, when an unknown
            printer took a galley of type and scrambled it to make a type
            specimen book.
          </div>
        </div>
        <div
          class="card-info"
        >
          <div
            class="author"
          >
            <strong>
              Authors: 
            </strong>
            Anna Sidorenko, Valentina Larina
          </div>
          <div
            class="duration"
          >
            Duration: 
            03:30
          </div>
          <div
            class="created"
          >
            Created: 
            10/11/2020
          </div>
          <div
            class="btn-wrap"
          >
            <button
              class="button"
            >
              Show course
            </button>
            <button
              class="button"
            >
              E
            </button>
            <button
              class="button"
            >
              D
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
